### DISCUSSION TITLE
Help with multiple smart objects and single folder input files

### DISCUSSION THREAD

**Discussion by zgdesign** [joonaspaakko/Batch-Mockup-Smart-Object-Replacement-photoshop-script#29] [Q&A] [answered]
Hello,
I just want to thank you once again for this amazing script. It's been a lifesaver and I use it all the time.
I currently have a situation and can't come up with a good solution.
I have mockups with 3 smart objects in each. I'd like to fill those with input images that are all in the same folder, but without duplicates, so basically 3 different images. The issue that I'm having is that each mockup fills all 3 smart objects with the same image. Is there any way to have the script advance to the next image from the same input folder after filling each smart object? I was hoping to avoid having 3 separate folders with different images if possible.
Thanks in advance for your help

**Comment by joonaspaakko** [comment, marked as answer]
> What a coincidence... Last Sunday I was approached about this same thing by email and I basically said "You can't really do it right now and it's too much of a niche scenario for me to implement". But by Monday I decided it wasn't such a bad idea and decided to add this feature. I didn't release it yet because I was bothered by how it leaves the previous loop's images in the last output file if there aren't enough input files. But since there seems to be demand for the feature, I decided to call it "experimental" in the changelog and push it online.

[v.1.9](https://github.com/joonaspaakko/Batch-Mockup-Smart-Object-Replacement-photoshop-script/blob/fc546baacf6ea29cab95eb3b97cb869ae2baeccd/script/Batch%20Mockup%20Smart%20Object%20Replacement.jsx) now has this feature and you use it by adding `input`, `inputNested`, and `inputFormats` to the smart object rather than each smart object:

```diff
#include "script/Batch Mockup Smart Object Replacement.jsx" 

var output = {
  path: './_output files', // Outputs next to the mockup psd
  format: 'jpg', // 'jpg', 'png', 'tif', 'psd', 'pdf'
  filename: '@mockup - $', 
};

mockups([
  
  // Mockup #1
  {
    output: output, 
    mockupPath: '$/assets', // Note: in v.1.9 you can also process multiple mockups with the same settings if you use a folder path here
    mockupNested: false,  // ...and process mockups in nested folders
+   input: [ '$/assets/_input files' ], 
+   inputNested: true,
    smartObjects: [
      
      // Smart object #1
      {
        target: '@screen',
-       input: [ '$/assets/_input files' ], 
-       inputNested: true,
        inputNested: true,
        align: 'center center',
        resize: 'fit', 
      },
      // Smart object #2
      {
        target: '@screen2',
-       input: [ '$/assets/_input files' ], 
-       inputNested: true,
        align: 'center center',
        resize: 'fill', 
      },
      
      // You could have more smart objects per mockup file...
      // {},
      
    ]
  },
  
  // You could have more mockup files...
  // {},
    
]);
```

**Reply by zgdesign** [reply]
Amazing, that works perfectly!
Thank you for your prompt replies and help as always!
Your work is invaluable and I hope other users appreciate it as much as I do :) 🥇

**Comment by zgdesign** [comment]
Hi,
I hope you're doing great, and happy upcoming 2024!!
I'm a bit stuck trying to figure out something that's probably simple. Basically I'm trying to get the previous functionality to work where all my smart objects are filled with the same artwork, versus different artwork within a folder.
I have 4 smart objects in my mockups named SO, SO1, SO2, and SO3. It's been working great getting them filled with 4 different input files, but now I can't figure out how to get all of them to be filled with the same artwork from the input folder.
I've tried setting inputNested: False, but that doesn't seem to work:

mockups([
  
  // Mockup #1
  {
        output: {
          path: '$/BATCH OUTPUT',
          format: 'jpg', // 'jpg', 'png', 'tif', 'psd', 'pdf'
          folders: false, // Files will be grouped by folders inside the output folder
           filename: '@mockup-@input', 
        },
        mockupPath: '$/Mockups_Final', 
        mockupNested: true,
        input: [ '../../../All_Design_Files'],  // Look for image files from this input folder...
        inputNested: false,
    smartObjects: [
      {
        target: 'SO',     
        resize: 'fit',
        inputNested: false,
      },
      {
          target: 'SO1',
          resize: 'fit',
          inputNested: false,
      },
      {
          target: 'SO2',
          resize: 'fit',
          inputNested: false,
        },
        {
            target: 'SO3', 
            resize: 'fit',
            inputNested: false,
        },
    ]
    },

**Reply by joonaspaakko** [reply]
I think what you need is to remove your current `input` property from the `mockup` object and instead copy the same exact property and path to each `smartobject` separately, much like all of those `inputNested` props. 

> You could also place the singular path in a variable, similar to how the output is in many of my examples, and then just copy and paste that variable in every smart object.

I have to say I'm not sure what you wanted to achieve with the `inputNested`. The only thing it does is look for input files from every single folder inside the given `input` path... So all it would do in this scenario (and in general) is increase the number of input files, if there are any suitable input files in a nested folder structure.

**Rambling:**

The way it works is admittedly somewhat illogical.

1. All `smartobject.input` work independently from each other. So if two of them for example share the same exact 'input` path, they're going to be placed separately on both... and in the same order as well. 
2. The `mockup.input` was added later to enable filling all smart objects sequentially from from a single pool of input files...
    - With this setup it's kinda paramount they all have the same exact input files, so the `input` was moved up a level to the `mockup` object and that's kinda why that is its only functionality. 
    - Since it was already possible to fill the same exact image to multiple or all smart object just by feeding them the same input paths separately, I didn't think to add that functionality to this as well. I also don't think it ever crossed my mind as a possible use-case
    - So in conclusion... It would've perhaps been more logical for `mockup.input` to not place input files sequentially by default, and rather have have an additional prop `mockup.inputSequentially` which would enable the behavior, but this is just kinda how the pieces landed this time.

**Reply by zgdesign** [reply]
Thanks for getting back to me with the detailed reply.
I tried removing the input property from the mockup object and copied the path to each smartobject separately, but now when I run the script Photoshop opens the first mockup file and then the script just stops.

**Reply by joonaspaakko** [reply]
I'm thinking it could be related to this issue: #30. I'm going to take a look at it later,  but FYI, the first thing I'll do is take the standalone example, add a new smart object in the psd, duplicate the settings script smart object and see what it does.

**Reply by zgdesign** [reply]
Ok, after a bit of testing I figured out what was causing the script to freeze/fail.
Instead of commenting out the input property from the mockup object, I had to leave it it in there but set it to an empty value. So instead of -->  //input: 'myPath',  I had to leave this in -->  input:'', 

But now I came across another issue and I'm not sure what caused it, however it happened several times. After running the script, I noticed that some of my layered mockup files were being flattened and saved that way. Luckily I have daily backups so I was able to restore the files, but I'm not sure what might have caused this to happen. During testing, I did stop the script short in Photoshop by hitting ESC. Any idea what might have caused this to happen?

Thanks again for all your help!

**Reply by joonaspaakko** [reply]
Hmmn... That sounds sounds like what it's supposed to do after placing the image inside the smart object; flatten and save... But it did it inside the mockup. Off the top of my head, it might be possible  this could happen  if it can't open the smart object, but thinks it did.

It might be safer to continue using 1.8.

**Reply by zgdesign** [reply]
What's odd is that I can't replicate the behaviour. I think it happened when I didn't have the parameters setup correctly, and when I had the input property commented out from the mockup object. Moving forward I'll make backups of the mockup files just in case this happens again, but otherwise it works great now.
Thank you!
